File: paper/tri_final_figures/make_drawio_export_preview.py
#!/usr/bin/env python3
"""Build a minimal localhost page that exports a draw.io file as PNG or SVG."""

from __future__ import annotations

import argparse
import json
from pathlib import Path


TEMPLATE = """<!doctype html>
<html><head><meta charset=\"utf-8\"><title>{title}</title>
<style>html,body{{margin:0;background:#fff}}iframe{{position:absolute;width:1px;height:1px;opacity:0}}#exported{{display:block;max-width:100%;height:auto;margin:0 auto}}</style>
</head><body><iframe id=\"drawio\" src=\"https://embed.diagrams.net/?embed=1&proto=json&spin=1&ui=min&libraries=0&grid=0&pv=0\"></iframe><img id=\"exported\" alt=\"draw.io export\"><script>
const xml={xml}; const format={fmt}; const scale={scale}; const frame=document.getElementById('drawio'); let sent=false;
function sendExport(){{frame.contentWindow.postMessage(JSON.stringify({{action:'export',format,xml,scale,border:0}}),'https://embed.diagrams.net')}}
window.addEventListener('message',e=>{{if(e.source!==frame.contentWindow||e.origin!=='https://embed.diagrams.net')return;let m=e.data;try{{if(typeof m==='string')m=JSON.parse(m)}}catch(_){{return}}if(!m)return;if(m.event==='export'&&m.data){{document.getElementById('exported').src=m.data;document.title='{title} - exported';return}}if(!sent&&(m.event==='init'||m.event==='configure')){{sent=true;sendExport()}}}});
setTimeout(()=>{{if(!sent){{sent=true;sendExport()}}}},2500);
</script></body></html>"""


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("drawio", type=Path)
    parser.add_argument("--out", type=Path, required=True)
    parser.add_argument("--format", choices=("png", "svg"), default="png")
    parser.add_argument("--scale", type=float, default=1.0)
    args = parser.parse_args()
    xml = args.drawio.read_text(encoding="utf-8")
    html = TEMPLATE.format(
        title=args.drawio.name,
        xml=json.dumps(xml),
        fmt=json.dumps(args.format),
        scale=json.dumps(args.scale),
    )
    args.out.write_text(html, encoding="utf-8")
    print(args.out)


if __name__ == "__main__":
    main()

File: paper/tri_final_figures/outputs/fig2_v18_paper_readable/make_drawio_export_preview.mjs
import fs from "node:fs/promises";
import path from "node:path";

const input = process.argv[2] || "fig2_tri_diagnostic_workflow_v18_paper_readable.drawio";
const output = process.argv[3] || "drawio-export-preview-v18.html";
const xml = await fs.readFile(input, "utf8");

const html = `<!doctype html>
<html>
<head>
<meta charset="utf-8" />
<title>TRI Figure 2 v18 canvas</title>
<style>
  html, body { margin: 0; min-height: 100%; background: #ffffff; }
  body { display: grid; place-items: center; padding: 20px; box-sizing: border-box; }
  #render { display: block; width: min(96vw, 1600px); height: auto; }
  #drawio { position: fixed; left: -10000px; top: -10000px; width: 960px; height: 400px; border: 0; }
  #status { font: 16px Arial, sans-serif; color: #264a56; }
</style>
</head>
<body>
<div id="status">Rendering editable draw.io source...</div>
<img id="render" alt="TRI Figure 2 v18" hidden />
<iframe id="drawio" src="https://embed.diagrams.net/?embed=1&proto=json&spin=1&ui=min&libraries=0&grid=0&pv=0"></iframe>
<script>
const xml = ${JSON.stringify(xml)};
const origin = "https://embed.diagrams.net";
const frame = document.getElementById("drawio");
const image = document.getElementById("render");
const status = document.getElementById("status");
let requested = false;

function requestExport() {
  if (requested) return;
  requested = true;
  frame.contentWindow.postMessage(JSON.stringify({
    action: "export",
    format: "png",
    xml,
    scale: 2,
    border: 0,
    background: "#ffffff"
  }), origin);
}

window.addEventListener("message", (event) => {
  if (event.source !== frame.contentWindow || event.origin !== origin) return;
  let message = event.data;
  try { if (typeof message === "string") message = JSON.parse(message); } catch { return; }
  if (!message) return;
  if (message.event === "init" || message.event === "configure") {
    requestExport();
  } else if (message.event === "export" && message.data) {
    image.src = message.data;
    image.hidden = false;
    status.remove();
  }
});

setTimeout(requestExport, 2500);
</script>
</body>
</html>`;

await fs.writeFile(path.resolve(output), html, "utf8");
